Avoid unnecessary reloading of point cloud when parameters like point size are updated

NEDJIMAbelgacem opened the pull request:
## Description
Previously if any parameter of a point cloud symbol is changed, the whole scene must be updated to reflect the changes.
For some parameters like point size, this is unnecessary. This PR adds the code to update the existing point cloud renderer instead or reseting the whole thing and redrawing the scene unnecessarily.
Here is a demo of how the scene updates instantaneously with reloading: 

https://user-images.githubusercontent.com/29183781/187203154-537d2707-c3b2-4976-a092-836bb6325b05.mp4

This currently works for point size and single color rendering.

NEDJIMAbelgacem committed:
Initial implementation of update behaviour

NEDJIMAbelgacem committed:
update renderer specific settings

benoitdm-oslandia commented on src/3d/qgspointcloudlayer3drenderer.cpp:
```diff
@@ -241,3 +251,42 @@ bool QgsPointCloudLayer3DRenderer::convertFrom2DRenderer( QgsPointCloudRenderer
   return true;
 }
+
+bool QgsPointCloudLayer3DRenderer::supportsUpdateFrom( QgsAbstract3DRenderer *renderer ) const
+{
+  QgsPointCloudLayer3DRenderer *r = dynamic_cast<QgsPointCloudLayer3DRenderer *>( renderer );
+  if ( !r || !r->symbol() )
+  {
+    return false;
+  }
+
+  const QgsPointCloud3DSymbol *rSymbol = r->symbol();
+  if ( rSymbol->symbolType() != mSymbol->symbolType() )
+  {
+    return false;
+  }
+
+  return true;
+}
+
+bool QgsPointCloudLayer3DRenderer::updateCurrentRenderer( QgsAbstract3DRenderer *renderer )
+{
+  Q_ASSERT( supportsUpdateFrom( renderer ) );
+  QgsPointCloudLayer3DRenderer *r = dynamic_cast<QgsPointCloudLayer3DRenderer *>( renderer );
+  setShowBoundingBoxes( r->showBoundingBoxes() );
+  bool updateScene = false;
+  if ( r->maximumScreenError() != maximumScreenError() )
+  {
+    setMaximumScreenError( r->maximumScreenError() );
+    updateScene = true;
+  }
+  if ( r->pointRenderingBudget() != pointRenderingBudget() )
+  {
+    setPointRenderingBudget( r->pointRenderingBudget() );
+    updateScene = true;
+  }
+  // TODO: fix const access
+  updateScene = updateScene || mSymbol->updateCurrentSymbol( const_cast< QgsPointCloud3DSymbol * >( r->symbol() ) );
+  return updateScene;
+}
```
Why did not you merge theses 2 functions? One should be enough?

```c++
bool QgsPointCloudLayer3DRenderer::updateCurrentRenderer( QgsAbstract3DRenderer *renderer )
{
  QgsPointCloudLayer3DRenderer *r = dynamic_cast<QgsPointCloudLayer3DRenderer *>( renderer );
  if ( !r || !r->symbol() )
  {
    return false;
  }

  const QgsPointCloud3DSymbol *rSymbol = r->symbol();
  if ( rSymbol->symbolType() != mSymbol->symbolType() )
  {
    return false;
  }
  if ( r->maximumScreenError() == maximumScreenError() || r->pointRenderingBudget() == pointRenderingBudget())
  {
    return false;
  }

  setShowBoundingBoxes( r->showBoundingBoxes() );
  setMaximumScreenError( r->maximumScreenError() );
  setPointRenderingBudget( r->pointRenderingBudget() );
  // TODO: fix const access
  updateScene = updateScene || mSymbol->updateCurrentSymbol( const_cast< QgsPointCloud3DSymbol * >( r->symbol() ) );
  return true;
}
```

NEDJIMAbelgacem replied:
For me it seems simpler to first check for updating support than to handle 3 the following cases in one go:
- The renderer is of same symbol as the new renderer but could be updated without resetting 
- The renderer is of same symbol as the new renderer but should be updated without resetting 
- The renderer is of a different symbol and needs to be completely reset

benoitdm-oslandia commented on src/core/3d/qgsabstract3drenderer.h:
```diff
@@ -64,6 +65,27 @@ class CORE_EXPORT QgsAbstract3DRenderer SIP_ABSTRACT
     //! Resolves references to other objects - second phase of loading - after readXml()
     virtual void resolveReferences( const QgsProject &project );
 
+    /**
+     * Returns whether the current renderer could be updated from \a renderer
+     * \since QGIS 3.28
+     */
+    virtual bool supportsUpdateFrom( QgsAbstract3DRenderer *renderer ) const
+    {
+      Q_UNUSED( renderer );
+      return false;
+    }
+
+    /**
+     * Updates the current renderer and returns true in case the 3D scene needs to be refreshed
+     * \warning The renderer \a renderer must be of the same type with the current renderer
+     * \sinec QGIS 3.28
+     */
+    virtual bool updateCurrentRenderer( QgsAbstract3DRenderer *renderer )
+    {
+      Q_UNUSED( renderer );
+      return false;
+    };
+
```
What about the other inherited classes (QgsMeshLayer3DRenderer, QgsVectorLayer3DRenderer and QgsRuleBased3DRenderer)?

NEDJIMAbelgacem replied:
The same logic will work for the other renderers: when the changed setting is only one QParameter passed to the shaders, change that parameter instead of updating the whole scene

benoitdm-oslandia reviewed (commented):
Nice job, it will speed up greatly the properties updates!

NEDJIMAbelgacem committed:
When the show bounding boxes is enabled,
the scene need to be moved for the bounding
boxes to appear..
A better solution is to just keep the bounding box
entities in memory and enable/disable it

NEDJIMAbelgacem committed:
fix QgsRgbPointCloud3DSymbol segfault

NEDJIMAbelgacem committed:
fix tests

nyalldawson commented on python/3d/auto_generated/symbols/qgspointcloud3dsymbol.sip.in:
```diff
@@ -155,6 +155,26 @@ than this threshold value will not be rendered, see :py:func:`~QgsPointCloud3DSy
+
+.. note::
+
+   This signal is only used to manage updating internal Qt3D QParameter object without reloading the whole scene
```
EIther this is a useful signal in general and should be exposed to Python (in which case the note isn't needed), or it's for internal use only and should be hidden from sip

nyalldawson commented on python/3d/auto_generated/symbols/qgspointcloud3dsymbol.sip.in:
```diff
@@ -220,7 +240,20 @@ Sets the color used by the renderer when using SingleColor rendering mode
+
+.. note::
+
+   This signal is only used to manage updating internal Qt3D QParameter object without reloading the whole scene
```
as above

nyalldawson commented on src/core/3d/qgsabstract3drenderer.h:
```diff
@@ -45,8 +46,9 @@ namespace Qt3DCore
-class CORE_EXPORT QgsAbstract3DRenderer SIP_ABSTRACT
+class CORE_EXPORT QgsAbstract3DRenderer : public QObject SIP_ABSTRACT
 {
+    Q_OBJECT
```
this shouldn't be a qobject

nyalldawson commented on src/core/3d/qgsabstract3dsymbol.h:
```diff
@@ -43,8 +43,9 @@ namespace Qt3DCore { class QEntity; } SIP_SKIP
  * \since QGIS 3.0
  */
-class CORE_EXPORT QgsAbstract3DSymbol
+class CORE_EXPORT QgsAbstract3DSymbol : public QObject
```
same here -- there's no reason to make this a qobject (and a ton of reasons not to)

nyalldawson commented on src/core/3d/qgsabstract3drenderer.h:
```diff
@@ -20,6 +20,7 @@
 #include "qgis_sip.h"
 
 #include <QString>
+#include <QDebug>
```
```suggestion
```

nyalldawson commented on src/3d/symbols/qgspointcloud3dsymbol.cpp:
```diff
@@ -25,16 +25,22 @@
-QgsPointCloud3DSymbol::QgsPointCloud3DSymbol()
+QgsPointCloud3DSymbol::QgsPointCloud3DSymbol( QgsPointCloud3DSymbol::RenderingStyle renderingStyle )
   : QgsAbstract3DSymbol()
+  , mRenderingStyle( renderingStyle )
```
This would be better handled as a pure virtual method in QgsPointCloud3DSymbol, rather than passing it to the constructor

nyalldawson commented on src/3d/qgspointcloudlayerchunkloader_p.h:
```diff
@@ -101,9 +101,9 @@ class QgsPointCloudLayerChunkLoader : public QgsChunkLoader
     /**
      * Constructs the loader
-     * QgsPointCloudLayerChunkLoader takes ownership over symbol
+     * \note since QGIS 3.28 the context doesn't take ownership over the passed symbol \a symbol
```
If we aren't cloning the symbol then we're going to run into all sorts of threading issues -- the symbols aren't thread safe and you'll get crashes if you modify the symbol in the main thread while the loader is doing stuff with them in a background thread.....

NEDJIMAbelgacem committed:
Address reviews 1

wonder-sk reviewed (commented):
As Nyall already suggested, we need a solution that does not make QgsPointCloudLayer3DRenderer and QgsPointCloud3DSymbol classes derived from QObject, and where QgsPointCloud3DSymbol objects are copied rather than shared to avoid possible threading issues...

NEDJIMAbelgacem committed:
fix concurrency concerns by making the renderer report the changed attributes and applying the changes to all entities
